### PR TITLE
VP-6823: AddDbContext without BuildServiceProvider()

### PR DESCRIPTION
**Pull request by eMazeika**

### CODE REVIEW

File: src/VirtoCommerce.OrdersModule.Web/Module.cs [hunk 1]
```diff
@@ -63,6 +65,7 @@ public void Initialize(IServiceCollection serviceCollection)
             serviceCollection.AddTransient<IAuthorizationHandler, OrderAuthorizationHandler>();
 
             serviceCollection.AddSingleton(typeof(IConverter), new SynchronizedConverter(new PdfTools()));
+            var configuration = serviceCollection.BuildServiceProvider().GetRequiredService<IConfiguration>();
```
Comment (tatarincev) [inline src/VirtoCommerce.OrdersModule.Web/Module.cs:68]: ???

Comment (eMazeika) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:68]: still unclear how to refactor:
serviceCollection.AddOptions<HtmlToPdfOptions>().Bind(**configuration**.GetSection("HtmlToPdf"))

**Review by akak1977 — DISMISSED** [review]
But HtmlToPdf init should be refactored, if possible.

File: src/VirtoCommerce.OrdersModule.Web/Module.cs (hunk 1 above)
Comment (tatarincev) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:68]: I think can avoid having to call `BuildServiceProvider` by registered the implementation of `IConfigureOptions<HtmlToPdfOptions>` interface that injected  `IConfiguration` as a dependency. 
See more details in this brilliant article.
https://andrewlock.net/accessing-services-when-configuring-mvcoptions-in-asp-net-core/#configureoptions-to-the-rescue-

Comment (eMazeika) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs]: the article is amazing

**Review by akak1977 — APPROVED** [review]

File: src/VirtoCommerce.OrdersModule.Web/Module.cs [hunk 2]
```diff
@@ -63,7 +66,7 @@ public void Initialize(IServiceCollection serviceCollection)
             serviceCollection.AddTransient<IAuthorizationHandler, OrderAuthorizationHandler>();
 
             serviceCollection.AddSingleton(typeof(IConverter), new SynchronizedConverter(new PdfTools()));
-            serviceCollection.AddOptions<HtmlToPdfOptions>().Bind(configuration.GetSection("HtmlToPdf")).ValidateDataAnnotations();
+            serviceCollection.AddSingleton<IConfigureOptions<HtmlToPdfOptions>, ConfigureHtmlToPdfOptions>();
```
Comment (avas) [inline src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: There is a much simpler way to bind options class to some configuration section without resolving `IConfiguration` - we just need to copy the [`BindConfiguration()` extension method](https://github.com/dotnet/runtime/blob/v5.0.2/src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/OptionsBuilderConfigurationExtensions.cs#L44-L74) from .NET Core 5 and place it in some shared assembly (e.g. in the platform or in `VirtoCommerce.CoreModule.Web.Common`). This way we will be able to retain the options registration code and use all of the existing `OptionsBuilder` extension methods, like this:
```c#
serviceCollection.AddOptions<HtmlToPdfOptions>()
    .BindConfiguration("HtmlToPdf")
    .ValidateDataAnnotations();
```

Comment (eMazeika) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: @tatarincev this ^^^ BindConfiguration() should be added to Platform, right?

Comment (OlegoO) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: @eMazeika  Create tech debt task for this

Comment (eMazeika) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: Created: https://virtocommerce.atlassian.net/browse/VP-6914

Comment (tatarincev) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: Indeed. Thanks @avas.  The resulting equivalent syntax will looks like: 
```C#
serviceCollection.AddOptions<HtmlToPdfOptions>().Configure<IConfiguration>((opts, config) =>
            {
                opts.Bind(config.GetSection("HtmlToPdf"));
            });
```
We don't need an extra extension method for this

Comment (avas) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: In this particular case this extension will help, but in general, IMO, this won't be enough - for example, you won't be able to add `.ValidateDataAnnotations()` or to call `.PostConfigure(...)`.

Comment (eMazeika) [reply src/VirtoCommerce.OrdersModule.Web/Module.cs:69]: The latest syntax:
```
            serviceCollection.AddOptions<HtmlToPdfOptions>().Configure<IConfiguration>((opts, config) =>
            {
                config.GetSection("HtmlToPdf").Bind(opts);
            });
```

**Pull request merged**